prevent-broker-scale-down-if-contains-partition-replicas.md

```diff
@@ -0,0 +1,58 @@
+
+## Motivation
+
+We should introduce the logic which can detect if the broker which is going to be removed still contains the partition replicas or not.
+If any partition replicas are still assigned to the broker which is going to be removed from the cluster, then we should get some warning in the status of the Kafka resource which will prompt users to do the reassignment and prevent the broker from being removed until the partition replicas are reassigned.
+
+## Proposal
+
+This proposal suggest how we can add the check to detect if the broker still contains any partition replicas and what to do if the broker scale down is not possible.
+
+## Implementation
+
+### Class
+
+A new utility class named `PreventBrokerScaleDownUtils` will be added to the assembly package of cluster operator module
+
+The reason for having this class is to keep all the utilities for checking the broker partition replicas separate such that we don't create a mess in the `KafkaReconciler` class by adding multiple large methods.
+Also, it will help us to test the changes easily.
+
+This class will have utility methods like `canScaleDownBrokers`, `canScaleDownBrokerCheck` and `brokerHasAnyReplicas` methods to check the if broker contains any partition replicas or not.
+
+This class will also contain methods like `topicNames`, `describeTopics` to query the topic metadata.
+
+#### Process:
+
+- When the broker count is changed in the Kafka resource, the `reconcile` method of the `KafkaReconciler` will be triggered to reconcile the Kafka brokers.
+- The `canScaleDownBrokers()` utility method will be present at the top of the compose chain in the `reconcile()` method of the `KafkaReconciler`.
+- This method will check if the broker contains any partition replicas or not and will continue the process based on the outcome.
+- For doing so, the topic metadata will be queried to detect if the broker contains any partition replicas.
+- An Admin client instance will be used to connect with the cluster and get us the topic details(topic name and topic description)
+- Then we can use this information to check if the broker contains any partition replicas or not.
+- The scale down is done after we make sure that the brokers that are going to be removed doesn't contain any partition replicas. By doing this we avoid any partial scale down.
+
+#### What to do if a broker contains partitions?
+
+#### Flow:
+
+- If partition replicas are found out on the broker we will revert back the kafka replicas to the previous count by changing the replica count directly in the kafka object and update the status with a warning stating that the scale down is not done and the replica count is changed back to previous one in the memory.
```

Comment:
1. What exactly are we changing it back _to_? The current value of the `spec.replicas` in the STS/SPS?
2. Be explicit whether or not the rest of the reconciliation carries on as normal (I think it does, right?)
3. Be explicit about the revised schema for the `status`. Is this situation modelled as a new condition in the `status.conditions`, or some other way?  In particular how does the user know what the current number of brokers in their cluster is? For example if you're saying the `Ready` condition will be present (because the rest of the reconciliation completed OK), and some other new condition (e.g. `CannotScaleDown`) is also present (to indicate that the scale down specifically can't be done right now), then the `spec.kafka.replicas` != the actual number of replicas, and the user can't know from the `Kafka` resource what the actual number of 
brokers in their cluster is.

Reply:
1. Yes. I am changing the `spec.replicas` in STS/SPS
2. Yes, it will run normally
3. I was planning to print the `spec.replicas` in the condition's message